test_client_no_client_cert fails

I'm running the test suite.

```
===================================== test session starts =====================================
platform linux -- Python 3.7.2, pytest-4.3.1, py-1.8.0, pluggy-0.9.0
cachedir: .tox/py37-extras/.pytest_cache
rootdir: /home/test/custodia/custodia, inifile: tox.ini
collected 162 items                                                                           

tests/test_authenticators.py ..
tests/test_cli.py .....
tests/test_custodia.py ..........................................F.
tests/test_httpd.py .
tests/test_ipa.py ........................
tests/test_message_kem.py ....
tests/test_misc.py ...
tests/test_plugins.py ....
tests/test_secrets.py ..................................
tests/test_server.py ....
tests/test_store.py ...
tests/test_store_sqlite.py ...........
tests/functional/test_basics.py ...
tests/functional/test_container.py .........
tests/functional/test_key.py ...........

========================================== FAILURES ===========================================
________________________ CustodiaHTTPSTests.test_client_no_client_cert ________________________

self = <test_custodia.CustodiaHTTPSTests testMethod=test_client_no_client_cert>

    def test_client_no_client_cert(self):
        client = CustodiaSimpleClient(self.socket_url + '/forwarder')
        client.headers['REMOTE_USER'] = 'test'
        client.set_ca_cert(self.ca_cert)
        # XXX workaround for requests bug with urllib3 v1.22
        with self.assertRaises(RequestsConnSSLErrors) as e:
            client.list_container('test')
        self.assert_ssl_error_msg("SSLV3_ALERT_HANDSHAKE_FAILURE",
>                                 e.exception)

tests/test_custodia.py:520: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_custodia.py:502: in assert_ssl_error_msg
    self.fail(str(exc))
E   AssertionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
```

It seems that urllib3 doesn't support TLS1.3 yet:
https://github.com/urllib3/urllib3/pull/1496

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,6 +1,6 @@
 sudo: required
 
-dist: trusty
+dist: xenial
 
 language: python
 
@@ -23,13 +23,17 @@ matrix:
       env: TOXENV=py36-extras
     - python: 3.6
       env: TOXENV=py36-noextras
-    - python: 3.6
+    - python: 3.7
+      env: TOXENV=py37-extras
+    - python: 3.7
+      env: TOXENV=py37-noextras
+    - python: 3.7
       env: TOXENV=doc
-    - python: 3.6
+    - python: 3.7
       env: TOXENV=lint
     - python: 2.7
       env: TOXENV=pep8py2
-    - python: 3.6
+    - python: 3.7
       env: TOXENV=pep8py3
 
 before_install:

--- a/Makefile
+++ b/Makefile
@@ -65,9 +65,9 @@ cscope:
 	git ls-files | xargs pycscope
 
 test: clean_socket clean_coverage
-	$(TOX) --skip-missing-interpreters -e py27-extra,py27-noextra
-	$(TOX) --skip-missing-interpreters -e py35-extra,py35-noextra
-	$(TOX) --skip-missing-interpreters -e py36-extra,py36-noextra
+	$(TOX) --skip-missing-interpreters -e py27-extras,py27-noextras
+	$(TOX) --skip-missing-interpreters -e py35-extras,py35-noextras
+	$(TOX) --skip-missing-interpreters -e py36-extras
 	$(TOX) --skip-missing-interpreters -e doc
 	$(TOX) -e coverage-report
 

--- a/src/custodia/cli/__init__.py
+++ b/src/custodia/cli/__init__.py
@@ -358,7 +358,7 @@ def main():
 
     try:
         result = args.func(args)
-    except BaseException as e:
+    except BaseException as e:  # pylint: disable=broad-except
         errcode, msg = error_message(args, e)
         main_parser.exit(errcode, msg)
     else:

--- a/src/custodia/ipa/interface.py
+++ b/src/custodia/ipa/interface.py
@@ -80,8 +80,8 @@ class IPAInterface(HTTPAuthenticator):
             # monkey-patch ipalib.constants and all loaded ipa modules
             ipalib.constants.USER_CACHE_PATH = rundir
             for name, mod in six.iteritems(sys.modules):
-                if (name.startswith(('ipalib.', 'ipaclient.')) and
-                        hasattr(mod, 'USER_CACHE_PATH')):
+                if (name.startswith(('ipalib.', 'ipaclient.'))
+                        and hasattr(mod, 'USER_CACHE_PATH')):
                     mod.USER_CACHE_PATH = rundir
 
         self._gssapi_config()

--- a/src/custodia/store/encgen.py
+++ b/src/custodia/store/encgen.py
@@ -48,8 +48,8 @@ class EncryptedOverlay(CSStore):
         self.store = None
         self.protected_header = None
 
-        if (not os.path.isfile(self.master_key) and
-                self.autogen_master_key):
+        if (not os.path.isfile(self.master_key)
+                and self.autogen_master_key):
             # XXX https://github.com/latchset/jwcrypto/issues/50
             size = self.key_sizes.get(self.master_enctype, 512)
             key = JWK(generate='oct', size=size)
@@ -79,8 +79,8 @@ class EncryptedOverlay(CSStore):
             if self.secret_protection == 'migrate':
                 self.set(key, value, replace=True)
             else:
-                raise CSStoreError('Secret Pinning check failed!' +
-                                   'Missing custodia.key element')
+                raise CSStoreError('Secret Pinning check failed!'
+                                   + 'Missing custodia.key element')
         elif jwe.jose_header['custodia.key'] != key:
             raise CSStoreError(
                 'Secret Pinning check failed! Expected {} got {}'.format(

--- a/src/custodia/store/sqlite.py
+++ b/src/custodia/store/sqlite.py
@@ -66,7 +66,7 @@ class SqliteStore(CSStore):
                 c.execute(setdata, (key, value))
         except sqlite3.IntegrityError as err:
             raise CSStoreExists(str(err))
-        except sqlite3.Error as err:
+        except sqlite3.Error:
             self.logger.exception("Error storing key %s", key)
             raise CSStoreError('Error occurred while trying to store key')
 

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -34,6 +34,16 @@ def pytest_addoption(parser):
 
 def pytest_runtest_setup(item):
     skip_servertest = item.config.getoption(SKIP_SERVERTEST)
-    if skip_servertest and item.get_marker("servertest") is not None:
+    skiptest = False
+    if skip_servertest:
+        # pytest < 4
+        if hasattr(item, 'get_marker'):
+            if item.get_marker("servertest"):
+                skiptest = True
+        # pytest >= 4
+        elif item.get_closest_marker("servertest"):
+            skiptest = True
+
+    if skiptest:
         # args has --skip-servertests and test is marked as servertest
         pytest.skip("Skip integration test")

--- a/tests/functional/base.py
+++ b/tests/functional/base.py
@@ -97,13 +97,11 @@ class CustodiaServerRunner(object):
     @pytest.fixture(scope="class")
     def custodia_server(self, simple_configuration, request, dev_null):
         # Don't write server messages to stdout unless we are in debug mode
-        # pylint: disable=no-member
-        if pytest.config.getoption('debug') or \
-                pytest.config.getoption('verbose'):
+        if (request.config.getoption('debug')
+                or request.config.getoption('verbose')):
             stdout = stderr = None
         else:
             stdout = stderr = dev_null
-        # pylint: enable=no-member
 
         self.process = subprocess.Popen(
             [sys.executable, '-m', 'custodia.server', self.custodia_conf],

--- a/tests/test_custodia.py
+++ b/tests/test_custodia.py
@@ -506,14 +506,16 @@ class CustodiaHTTPSTests(CustodiaTests):
         client.set_client_cert(cls.client_cert, cls.client_key)
         return client
 
-    def assert_ssl_error_msg(self, msg, exc):
-        # CERTIFICATE_VERIFY_FAILED, SSLV3_ALERT_HANDSHAKE_FAILURE
-        if msg in str(exc):
-            return
-        # 'certificate verify failed'
-        msg = msg.lower().replace('_', ' ')
-        if msg in str(exc):
-            return
+    def assert_ssl_error_msg(self, msgs, exc):
+        # CERTIFICATE_VERIFY_FAILED or
+        # SSLV3_ALERT_HANDSHAKE_FAILURE / Connection reset by peer
+        for msg in msgs:
+            if msg in str(exc):
+                return
+            # 'certificate verify failed'
+            msg = msg.lower().replace('_', ' ')
+            if msg in str(exc):
+                return
         self.fail(str(exc))
 
     def test_client_no_ca_trust(self):
@@ -522,7 +524,7 @@ class CustodiaHTTPSTests(CustodiaTests):
         # XXX workaround for requests bug with urllib3 v1.22
         with self.assertRaises(RequestsConnSSLErrors) as e:
             client.list_container('test')
-        self.assert_ssl_error_msg("CERTIFICATE_VERIFY_FAILED", e.exception)
+        self.assert_ssl_error_msg(["CERTIFICATE_VERIFY_FAILED"], e.exception)
 
     def test_client_no_client_cert(self):
         client = CustodiaSimpleClient(self.socket_url + '/forwarder')
@@ -531,7 +533,8 @@ class CustodiaHTTPSTests(CustodiaTests):
         # XXX workaround for requests bug with urllib3 v1.22
         with self.assertRaises(RequestsConnSSLErrors) as e:
             client.list_container('test')
-        self.assert_ssl_error_msg("SSLV3_ALERT_HANDSHAKE_FAILURE",
+        self.assert_ssl_error_msg(["Connection reset by peer",
+                                   "SSLV3_ALERT_HANDSHAKE_FAILURE"],
                                   e.exception)
 
     def test_C_client_cert_auth(self):

--- a/tox.ini
+++ b/tox.ini
@@ -1,6 +1,6 @@
 [tox]
 minversion = 2.3.1
-envlist = lint,py{27,35,36}-{extras,noextras},pep8py2,pep8py3,doc,coverage-report
+envlist = lint,py{27,35,36,37}-{extras,noextras},pep8py2,pep8py3,doc,coverage-report
 skip_missing_interpreters = true
 
 [testenv]

--- a/tox.ini
+++ b/tox.ini
@@ -90,5 +90,5 @@ addopts=-p no:logging
 exclude = .tox,*.egg,dist,build,docs/source
 show-source = true
 max-line-length = 79
-ignore = N802
+ignore = N802,W503
 application-import-names = custodia

